GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1094,14 +1099,25 @@ private void UpdatePlaceholders()
 
                 this.blobSizes.Flush();
 
-                // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
-                // so that we don't try to delete a folder placeholder that has file placeholders and just fails
-                foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderListCopy.Where(x => x.IsFolder).OrderByDescending(x => x.Path))
-                {
-                    this.TryRemoveFolderPlaceholder(folderPlaceholder, updatedPlaceholderList, folderPlaceholdersToKeep);
+                using (BlobSizes.BlobSizesConnection blobSizesConnection = this.blobSizes.CreateConnection())
+                {
+                    // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
+                    // so that we don't try to delete a folder placeholder that has file placeholders and just fails
+                    IEnumerable<PlaceholderListDatabase.PlaceholderData> folderPlaceholderData = placeholderListCopy.Where(x => x.IsFolder);
+
+                    HashSet<string> folderPlaceholders = new HashSet<string>(folderPlaceholderData.Select(x => x.Path), StringComparer.OrdinalIgnoreCase);
```

Comment:
Let me measure that as well.

One question I had looking at this code, do you remember why the folder placeholder list is ordered by descending before calling TryRemoveFolderPlaceholder? 

I assumed it was being done so that child folders would be removed before their parents, does that sound right to you?

>  is now being enumerated twice. Here and the line below

Another option I'll look into is removing this call (higher up in the method):

``` placeholderListCopy.Where(x => !x.IsFolder).ToList(),```

And replacing it with code that will walk `placeholderListCopy` to build both a file list and the `HashSet`.

Reply:
Yes, we have to make sure all the children are removed first so that we can remove the folder when it is empty since we are only removing individual folders and are depending on getting the directory not empty error if we can't delete because of existing files or folders.

Reply:
Do we have tests that fail if we drop that ordering? This is a very subtle detail that is easy to lose track of over time.

Reply:
> Do we have tests that fail if we drop that ordering? This is a very subtle detail that is easy to lose track of over time.

I'll try taking that out and see if any tests fail.  The tricky part is even if the ordering were to be removed, I'm not sure that we can come up with a test that would be guaranteed to catch it (as whatever random ordering VFSForGit would use in the future might just happen to work with the test).